File: BPLS.go
package main

// cd "C:\Users\antho\Programing Projects"; go build -o bpls.exe BPLS.go; .\bpls.exe
import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

func clear() {
	cmd := exec.Command("clear") // Linux / Mac
	if os.Getenv("OS") == "Windows_NT" {
		cmd = exec.Command("cmd", "/c", "cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

type BPLSInterpreter struct {
	objects   map[string]string
	groups    map[string][]string
	variables map[string]string
	queue     []string
}

func NewBPLSInterpreter() *BPLSInterpreter {
	return &BPLSInterpreter{
		objects:   make(map[string]string),
		groups:    make(map[string][]string),
		variables: make(map[string]string),
		queue:     []string{},
	}
}

func (b *BPLSInterpreter) substitute(token string) string {
	if val, ok := b.variables[token]; ok {
		return val
	}
	return token
}

func indexOf(slice []string, val string) int {
	for i, v := range slice {
		if v == val {
			return i
		}
	}
	return -1
}

func contains(slice []string, val string) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func removeFromSlice(slice []string, val string) []string {
	result := []string{}
	for _, v := range slice {
		if v != val {
			result = append(result, v)
		}
	}
	return result
}

func splitArgs(input string) []string {
	var args []string
	var current strings.Builder
	inQuotes := false
	for _, r := range input {
		switch r {
		case ' ':
			if inQuotes {
				current.WriteRune(r)
			} else if current.Len() > 0 {
				args = append(args, current.String())
				current.Reset()
			}
		case '"':
			inQuotes = !inQuotes
		default:
			current.WriteRune(r)
		}
	}
	if current.Len() > 0 {
		args = append(args, current.String())
	}
	return args
}

func (b *BPLSInterpreter) Run(input string) string {
	tokens := splitArgs(input)
	if len(tokens) == 0 {
		return "Error: Empty command."
	}

	cmd := strings.ToUpper(tokens[0])

	switch cmd {
	case "VARIABLE":
		if len(tokens) >= 4 && strings.ToUpper(tokens[2]) == "IS" {
			varName := strings.Trim(tokens[1], `"`)
			value := strings.Trim(tokens[3], `"`)
			b.variables[varName] = value
			return fmt.Sprintf("Variable '%s' set to '%s'.", varName, value)
		}

	case "CREATE":
		if len(tokens) >= 3 {
			switch strings.ToUpper(tokens[1]) {
			case "OBJ":
				objName := strings.Trim(tokens[2], `"`)
				b.objects[objName] = ""
				return fmt.Sprintf("Object '%s' created.", objName)
			case "GROUP":
				groupName := strings.Trim(tokens[2], `"`)
				b.groups[groupName] = []string{}
				return fmt.Sprintf("Group '%s' created.", groupName)
			}
		}

	case "CLEAR":
		clear()
		return ""

	case "LOC":
		if len(tokens) >= 2 {
			obj := b.substitute(tokens[1])
			group, ok := b.objects[obj]
			if !ok {
				return fmt.Sprintf("Error: Object '%s' not found.", obj)
			}
			if group == "" {
				return fmt.Sprintf("Object '%s' is not in any group.", obj)
			}
			idx := indexOf(b.groups[group], obj)
			return fmt.Sprintf("Object '%s' is located in group '%s' at ITEM %d.", obj, group, idx+1)
		}

	case "FIND":
		if len(tokens) >= 5 && strings.ToUpper(tokens[1]) == "ITEM" && strings.ToUpper(tokens[3]) == "OF" {
			n, err := strconv.Atoi(tokens[2])
			if err != nil {
				return "Error: ITEM number must be an integer."
			}
			group := b.substitute(tokens[4])
			list, ok := b.groups[group]
			if !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			if n <= 0 || n > len(list) {
				return fmt.Sprintf("Error: Group '%s' does not have ITEM %d.", group, n)
			}
			return fmt.Sprintf("Group '%s' ITEM %d is '%s'.", group, n, list[n-1])
		}

	case "SUM":
		if len(tokens) >= 2 {
			group := b.substitute(tokens[1])
			list, ok := b.groups[group]
			if !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			sum := 0
			count := 0
			for _, v := range list {
				if n, err := strconv.Atoi(v); err == nil {
					sum += n
					count++
				}
			}
			if count == 0 {
				return fmt.Sprintf("No integers in group '%s'.", group)
			}
			return fmt.Sprintf("SUM of '%s' = %d", group, sum)
		}

	case "AVG":
		if len(tokens) >= 2 {
			group := b.substitute(tokens[1])
			list, ok := b.groups[group]
			if !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			sum := 0
			count := 0
			for _, v := range list {
				if n, err := strconv.Atoi(v); err == nil {
					sum += n
					count++
				}
			}
			if count == 0 {
				return fmt.Sprintf("No integers in group '%s'.", group)
			}
			return fmt.Sprintf("AVG of '%s' = %f", group, float64(sum)/float64(count))
		}

	case "SWAP":
		upperTokens := make([]string, len(tokens))
		for i, t := range tokens {
			upperTokens[i] = strings.ToUpper(t)
		}
		if contains(upperTokens, "WITH") && contains(upperTokens, "IN") {
			withIdx := indexOf(upperTokens, "WITH")
			inIdx := indexOf(upperTokens, "IN")
			if withIdx < 1 || inIdx <= withIdx {
				return "Error: Invalid SWAP syntax."
			}
			obj1 := tokens[1]
			obj2 := tokens[withIdx+1]
			group := b.substitute(tokens[inIdx+1])
			list, ok := b.groups[group]
			if !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			for i, v := range list {
				if v == obj1 {
					list[i] = obj2
				} else if v == obj2 {
					list[i] = obj1
				}
			}
			b.groups[group] = list
			return fmt.Sprintf("Swapped '%s' with '%s' in group '%s'.", obj1, obj2, group)
		}

	case "MOVE":
		if len(tokens) >= 5 && strings.ToUpper(tokens[1]) == "VARIABLE" && strings.ToUpper(tokens[3]) == "TO" {
			varName := strings.Trim(tokens[2], `"`)
			objName, ok := b.variables[varName]
			if !ok {
				return fmt.Sprintf("Error: Variable '%s' not found.", varName)
			}
			group := strings.Trim(tokens[4], `"`)
			if _, ok := b.groups[group]; !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			if _, ok := b.objects[objName]; !ok {
				return fmt.Sprintf("Error: Object '%s' not found.", objName)
			}
			oldGroup := b.objects[objName]
			if oldGroup != "" {
				b.groups[oldGroup] = removeFromSlice(b.groups[oldGroup], objName)
			}
			b.groups[group] = append(b.groups[group], objName)
			b.objects[objName] = group
			return fmt.Sprintf("Object '%s' (from variable '%s') moved to group '%s'.", objName, varName, group)
		}
		if len(tokens) >= 3 && strings.ToUpper(tokens[2]) == "TO" {
			objName := strings.Trim(tokens[1], `"`)
			group := strings.Trim(tokens[3], `"`)
			if _, ok := b.groups[group]; !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			if _, ok := b.objects[objName]; !ok {
				return fmt.Sprintf("Error: Object '%s' not found.", objName)
			}
			oldGroup := b.objects[objName]
			if oldGroup != "" {
				b.groups[oldGroup] = removeFromSlice(b.groups[oldGroup], objName)
			}
			b.groups[group] = append(b.groups[group], objName)
			b.objects[objName] = group
			return fmt.Sprintf("Object '%s' moved to group '%s'.", objName, group)
		}

	case "REMOVE":
		if len(tokens) >= 5 && strings.ToUpper(tokens[1]) == "OBJ" && strings.ToUpper(tokens[3]) == "FROM" {
			objName := strings.Trim(tokens[2], `"`)
			group := strings.Trim(tokens[4], `"`)
			list, ok := b.groups[group]
			if !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			if !contains(list, objName) {
				return fmt.Sprintf("Error: Object '%s' not in group '%s'.", objName, group)
			}
			b.groups[group] = removeFromSlice(list, objName)
			b.objects[objName] = ""
			return fmt.Sprintf("Object '%s' removed from group '%s'.", objName, group)
		}

	case "LIST":
		if len(tokens) >= 2 {
			group := b.substitute(tokens[1])
			list, ok := b.groups[group]
			if !ok {
				return fmt.Sprintf("Error: Group '%s' not found.", group)
			}
			if len(list) == 0 {
				return fmt.Sprintf("Group '%s' is empty.", group)
			}
			return fmt.Sprintf("Group '%s': %s", group, strings.Join(list, ", "))
		}

	case "SAVE":
		if len(tokens) >= 4 && strings.ToUpper(tokens[1]) == "CODE" && strings.ToUpper(tokens[2]) == "TO" {
			filename := tokens[3]
			f, err := os.Create(filename)
			if err != nil {
				return fmt.Sprintf("Error creating file '%s': %v", filename, err)
			}
			defer f.Close()
			data := map[string]interface{}{
				"objects":   b.objects,
				"groups":    b.groups,
				"variables": b.variables,
			}
			enc := json.NewEncoder(f)
			if err := enc.Encode(data); err != nil {
				return fmt.Sprintf("Error encoding JSON: %v", err)
			}
			return fmt.Sprintf("Code saved to '%s'.", filename)
		}

	case "LOAD":
		if len(tokens) >= 2 {
			filename := tokens[1]
			f, err := os.Open(filename)
			if err != nil {
				return fmt.Sprintf("Error opening file '%s': %v", filename, err)
			}
			defer f.Close()
			var data map[string]interface{}
			dec := json.NewDecoder(f)
			if err := dec.Decode(&data); err != nil {
				return fmt.Sprintf("Error decoding JSON: %v", err)
			}
			if objs, ok := data["objects"].(map[string]interface{}); ok {
				b.objects = make(map[string]string)
				for k, v := range objs {
					b.objects[k] = v.(string)
				}
			}
			if grps, ok := data["groups"].(map[string]interface{}); ok {
				b.groups = make(map[string][]string)
				for k, v := range grps {
					arr := []string{}
					if slice, ok := v.([]interface{}); ok {
						for _, s := range slice {
							arr = append(arr, s.(string))
						}
					}
					b.groups[k] = arr
				}
			}
			if vars, ok := data["variables"].(map[string]interface{}); ok {
				b.variables = make(map[string]string)
				for k, v := range vars {
					b.variables[k] = v.(string)
				}
			}
			return fmt.Sprintf("Code loaded from '%s'.", filename)
		}
	}

	return "Error: Command not recognized or incomplete."
}

func main() {
	clear()
	interpreter := NewBPLSInterpreter()
	scanner := bufio.NewScanner(os.Stdin)

	fmt.Println("BPLS v1.3 - Beginner’s Programming Language for Statistics")
	fmt.Println("Type commands to queue them. Use RUN CODE to execute all queued commands.")
	fmt.Println("Type EXIT or QUIT to leave.\n")

	for {
		fmt.Print("BPLS> ")
		if !scanner.Scan() {
			fmt.Println("\nExiting BPLS interpreter.")
			break
		}
		cmd := scanner.Text()
		upperCmd := strings.ToUpper(strings.TrimSpace(cmd))

		if upperCmd == "EXIT" || upperCmd == "QUIT" {
			fmt.Println("Exiting BPLS interpreter.")
			break
		}

		if upperCmd == "RUN CODE" {
			start := time.Now()
			fmt.Printf("START: %v\n", start.Format("15:04:05"))

			for _, line := range interpreter.queue {
				output := interpreter.Run(line)
				if output != "" {
					fmt.Println(output)
				}
			}

			end := time.Now()
			diff := end.Sub(start)
			fmt.Printf("END:   %v\n", end.Format("15:04:05"))
			fmt.Printf("DURATION: %v\n", diff)

			interpreter.queue = []string{} // clear queue
		} else {
			interpreter.queue = append(interpreter.queue, cmd)
		}
	}
}

File: BPLS.py
import os
import json
import shlex
import time

def clear():
    os.system('cls' if os.name == 'nt' else 'clear')

class BPLSInterpreter:
    def __init__(self):
        self.objects = {}     # objects
        self.groups = {}      # groups
        self.variables = {}   # variables
        self.queue = []       # queued commands

    def substitute(self, token):
        return self.variables.get(token, token)

    def index_of(self, lst, val):
        try:
            return lst.index(val)
        except ValueError:
            return -1

    def remove_from_list(self, lst, val):
        return [x for x in lst if x != val]

    def split_args(self, input_str):
        return shlex.split(input_str)

    def run(self, input_str):
        tokens = self.split_args(input_str)
        if not tokens:
            return "Error: Empty command."

        cmd = tokens[0].upper()

        # VARIABLE "varname" IS "value"
        if cmd == "VARIABLE" and len(tokens) >= 4 and tokens[2].upper() == "IS":
            varname = tokens[1].strip('"')
            value = tokens[3].strip('"')
            self.variables[varname] = value
            return f"Variable '{varname}' set to '{value}'."

        # CREATE OBJ / GROUP
        elif cmd == "CREATE" and len(tokens) >= 3:
            if tokens[1].upper() == "OBJ":
                objname = tokens[2].strip('"')
                self.objects[objname] = ""
                return f"Object '{objname}' created."
            elif tokens[1].upper() == "GROUP":
                groupname = tokens[2].strip('"')
                self.groups[groupname] = []
                return f"Group '{groupname}' created."

        elif cmd == "CLEAR":
            clear()
            return ""

        # LOC
        elif cmd == "LOC" and len(tokens) >= 2:
            obj = self.substitute(tokens[1])
            if obj not in self.objects:
                return f"Error: Object '{obj}' not found."
            group = self.objects[obj]
            if group == "":
                return f"Object '{obj}' is not in any group."
            idx = self.index_of(self.groups[group], obj)
            return f"Object '{obj}' is located in group '{group}' at ITEM {idx+1}."

        # FIND ITEM
        elif cmd == "FIND" and len(tokens) >= 5 and tokens[1].upper() == "ITEM" and tokens[3].upper() == "OF":
            try:
                n = int(tokens[2])
            except ValueError:
                return "Error: ITEM number must be an integer."
            group = self.substitute(tokens[4])
            if group not in self.groups:
                return f"Error: Group '{group}' not found."
            lst = self.groups[group]
            if n <= 0 or n > len(lst):
                return f"Error: Group '{group}' does not have ITEM {n}."
            return f"Group '{group}' ITEM {n} is '{lst[n-1]}'."

        # SUM / AVG
        elif cmd == "SUM" and len(tokens) >= 2:
            group = self.substitute(tokens[1])
            if group not in self.groups:
                return f"Error: Group '{group}' not found."
            ints = [int(v) for v in self.groups[group] if v.isdigit()]
            if not ints:
                return f"No integers in group '{group}'."
            return f"SUM of '{group}' = {sum(ints)}"

        elif cmd == "AVG" and len(tokens) >= 2:
            group = self.substitute(tokens[1])
            if group not in self.groups:
                return f"Error: Group '{group}' not found."
            ints = [int(v) for v in self.groups[group] if v.isdigit()]
            if not ints:
                return f"No integers in group '{group}'."
            return f"AVG of '{group}' = {sum(ints)/len(ints)}"

        # SWAP
        elif cmd == "SWAP":
            if "WITH" in [t.upper() for t in tokens] and "IN" in [t.upper() for t in tokens]:
                with_idx = [t.upper() for t in tokens].index("WITH")
                in_idx = [t.upper() for t in tokens].index("IN")
                obj1 = tokens[1]
                obj2 = tokens[with_idx+1]
                group = self.substitute(tokens[in_idx+1])
                if group not in self.groups:
                    return f"Error: Group '{group}' not found."
                self.groups[group] = [obj2 if v == obj1 else obj1 if v == obj2 else v for v in self.groups[group]]
                return f"Swapped '{obj1}' with '{obj2}' in group '{group}'."

        # MOVE VARIABLE TO / MOVE OBJ TO
        elif cmd == "MOVE":
            if len(tokens) >= 5 and tokens[1].upper() == "VARIABLE" and tokens[3].upper() == "TO":
                varname = tokens[2].strip('"')
                if varname not in self.variables:
                    return f"Error: Variable '{varname}' not found."
                objname = self.variables[varname]
                group = tokens[4].strip('"')
                if group not in self.groups:
                    return f"Error: Group '{group}' not found."
                if objname not in self.objects:
                    return f"Error: Object '{objname}' not found."
                old_group = self.objects[objname]
                if old_group:
                    self.groups[old_group] = self.remove_from_list(self.groups[old_group], objname)
                self.groups[group].append(objname)
                self.objects[objname] = group
                return f"Object '{objname}' (from variable '{varname}') moved to group '{group}'."
            elif len(tokens) >= 4 and tokens[2].upper() == "TO":
                objname = tokens[1].strip('"')
                group = tokens[3].strip('"')
                if group not in self.groups:
                    return f"Error: Group '{group}' not found."
                if objname not in self.objects:
                    return f"Error: Object '{objname}' not found."
                old_group = self.objects[objname]
                if old_group:
                    self.groups[old_group] = self.remove_from_list(self.groups[old_group], objname)
                self.groups[group].append(objname)
                self.objects[objname] = group
                return f"Object '{objname}' moved to group '{group}'."

        # REMOVE OBJ FROM GROUP
        elif cmd == "REMOVE" and len(tokens) >= 5 and tokens[1].upper() == "OBJ" and tokens[3].upper() == "FROM":
            objname = tokens[2].strip('"')
            group = tokens[4].strip('"')
            if group not in self.groups:
                return f"Error: Group '{group}' not found."
            if objname not in self.groups[group]:
                return f"Error: Object '{objname}' not in group '{group}'."
            self.groups[group] = self.remove_from_list(self.groups[group], objname)
            self.objects[objname] = ""
            return f"Object '{objname}' removed from group '{group}'."

        # DELETE OBJ / GROUP
        elif cmd == "DELETE":
            if len(tokens) >= 3 and tokens[1].upper() == "OBJ":
                objname = tokens[2].strip('"')
                if objname not in self.objects:
                    return f"Error: Object '{objname}' not found."
                group = self.objects[objname]
                if group and objname in self.groups[group]:
                    self.groups[group] = self.remove_from_list(self.groups[group], objname)
                del self.objects[objname]
                return f"Object '{objname}' deleted."
            elif len(tokens) >= 3 and tokens[1].upper() == "GROUP":
                group = tokens[2].strip('"')
                if group not in self.groups:
                    return f"Error: Group '{group}' not found."
                for obj in self.groups[group]:
                    self.objects[obj] = ""
                del self.groups[group]
                return f"Group '{group}' deleted."

        # LIST
        elif cmd == "LIST" and len(tokens) >= 2:
            group = self.substitute(tokens[1])
            if group not in self.groups:
                return f"Error: Group '{group}' not found."
            lst = self.groups[group]
            if not lst:
                return f"Group '{group}' is empty."
            return f"Group '{group}': {', '.join(lst)}"

        # SAVE / LOAD
        elif cmd == "SAVE" and len(tokens) >= 4 and tokens[1].upper() == "CODE" and tokens[2].upper() == "TO":
            filename = tokens[3]
            snapshot = {"objects": self.objects, "groups": self.groups, "variables": self.variables}
            with open(filename, "w") as f:
                json.dump(snapshot, f)
            return f"Group '{group}': {', '.join(lst)}"

        # SAVE CODE TO <filename>
        elif cmd == "SAVE" and len(tokens) >= 4 and tokens[1].upper() == "CODE" and tokens[2].upper() == "TO":
            filename = tokens[3]
            snapshot = {"objects": self.objects, "groups": self.groups, "variables": self.variables}
            try:
                with open(filename, "w") as f:
                    json.dump(snapshot, f)
                return f"Code saved to '{filename}'."
            except Exception as e:
                return f"Error saving code: {e}"

        # LOAD <filename>
        elif cmd == "LOAD" and len(tokens) >= 2:
            filename = tokens[1]
            if not os.path.exists(filename):
                return f"Error: File '{filename}' not found."
            try:
                with open(filename, "r") as f:
                    snapshot = json.load(f)
                self.objects = snapshot.get("objects", {})
                self.groups = snapshot.get("groups", {})
                self.variables = snapshot.get("variables", {})
                return f"Code loaded from '{filename}'."
            except Exception as e:
                return f"Error loading code: {e}"

        # MAKE FILE <file_name>
        elif cmd == "MAKE" and len(tokens) >= 3 and tokens[1].upper() == "FILE":
            filename = tokens[2].strip('"')
            try:
                with open(filename, "w"):
                    pass
                return f"File '{filename}' created."
            except Exception as e:
                return f"Error creating file '{filename}': {e}"

        # MAKE FOLDER <folder_name>
        elif cmd == "MAKE" and len(tokens) >= 3 and tokens[1].upper() == "FOLDER":
            foldername = tokens[2].strip('"')
            try:
                os.makedirs(foldername, exist_ok=True)
                return f"Folder '{foldername}' created."
            except Exception as e:
                return f"Error creating folder '{foldername}': {e}"

        # PUT "group_name" IN "file_name"
        elif cmd == "PUT" and len(tokens) >= 4 and tokens[2].upper() == "IN":
            group = self.substitute(tokens[1].strip('"'))
            filename = tokens[3].strip('"')
            if group not in self.groups:
                return f"Error: Group '{group}' not found."
            try:
                with open(filename, "w") as f:
                    for obj in self.groups[group]:
                        f.write(obj + "\n")
                return f"Group '{group}' written into file '{filename}'."
            except Exception as e:
                return f"Error writing group '{group}' to file '{filename}': {e}"

        return "Error: Command not recognized or incomplete."


def main():
    clear()
    interpreter = BPLSInterpreter()
    print("BPLS v1.3 - Beginner’s Programming Language for Statistics")
    print("Type commands to queue them. Use RUN CODE to execute all queued commands.")
    print("Type EXIT or QUIT to leave.\n")

    while True:
        try:
            cmd = input("BPLS> ").strip()
            upper_cmd = cmd.upper()
            if upper_cmd in ["EXIT", "QUIT"]:
                print("Exiting BPLS interpreter.")
                break
            elif upper_cmd == "RUN CODE":
                start = time.time()
                print(f"START: {time.strftime('%H:%M:%S', time.localtime(start))}")
                for line in interpreter.queue:
                    output = interpreter.run(line)
                    if output:
                        print(output)
                end = time.time()
                print(f"END:   {time.strftime('%H:%M:%S', time.localtime(end))}")
                print(f"DURATION: {end - start:.2f} seconds")
                interpreter.queue = []
            else:
                interpreter.queue.append(cmd)
        except KeyboardInterrupt:
            print("\nExiting BPLS interpreter.")
            break


if __name__ == "__main__":
    main()
